server/services/v1/observability.go

```diff
@@ -0,0 +1,155 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"strconv"
+
+	"github.com/fullstorydev/grpchan/inprocgrpc"
+	"github.com/go-chi/chi/v5"
+	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
+	"github.com/rs/zerolog/log"
+	api "github.com/tigrisdata/tigris/api/server/v1"
+	"github.com/tigrisdata/tigris/server/config"
+	"google.golang.org/grpc"
+)
+
+const (
+	observabilityPattern     = "/observability/*"
+	AcceptHeader             = "Accept"
+	ApplicationJsonHeaderVal = "application/json"
+	DDApiKey                 = "DD-API-KEY"
+	DDAppKey                 = "DD-APPLICATION-KEY"
+	Query                    = "query"
+)
+
+type observabilityService struct {
+	api.UnimplementedObservabilityServer
+	ObservableProvider
+}
+
+type ObservableProvider interface {
+	QueryTimeSeriesMetrics(ctx context.Context, request *api.QueryTimeSeriesMetricsRequest) (*api.QueryTimeSeriesMetricsResponse, error)
+}
+
+type Datadog struct {
+	config.ObservabilityConfig
+}
+
+type MetricsQueryResp struct {
+	FromDate int64 `json:"from_date"`
+	ToDate   int64 `json:"to_date"`
+	Query    string
+	Series   []struct {
+		PointList [][]float64 `json:"pointlist"`
+	}
+}
+
+func (dd Datadog) QueryTimeSeriesMetrics(_ context.Context, req *api.QueryTimeSeriesMetricsRequest) (*api.QueryTimeSeriesMetricsResponse, error) {
+	ddReq, err := http.NewRequest(http.MethodGet, dd.ObservabilityConfig.ProviderUrl+"/api/v1/query", nil)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to query metrics: reason = "+err.Error())
+	}
+
+	q := ddReq.URL.Query()
+	q.Add("from", strconv.FormatInt(req.From, 10))
+	q.Add("to", strconv.FormatInt(req.To, 10))
+	if config.DefaultConfig.Auth.EnableNamespaceIsolation {
+		// TODO: add tag to scope query to the right namespace
+		q.Add("query", req.Query)
+	} else {
+		q.Add("query", req.Query)
+	}
+	ddReq.URL.RawQuery = q.Encode()
+	ddReq.Header.Add(AcceptHeader, ApplicationJsonHeaderVal)
+	ddReq.Header.Add(DDApiKey, config.DefaultConfig.Observability.ApiKey)
+	ddReq.Header.Add(DDAppKey, config.DefaultConfig.Observability.AppKey)
+	httpClient := &http.Client{}
+
+	resp, err := httpClient.Do(ddReq)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to query metrics: reason = "+err.Error())
+	}
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to query metrics: reason = "+err.Error())
+	}
+
+	bodyStr := string(body)
+	if resp.StatusCode == 200 {
+		var ddResp MetricsQueryResp
+		err = json.Unmarshal(body, &ddResp)
+		var dataPoints = make([]*api.DataPoint, len(ddResp.Series[0].PointList))
```

Comment:
check that there is atleast one series in the response before accessing it.

Reply:
added check.